### PR TITLE
[MB-7489] Custom eslint plugin for CATO

### PR DESCRIPTION
**Pull request by Ronolibert**
## Description

This adds a custom eslint plugin to the project that checks for annotations when disabling certain eslint rules that may have ATO implications. This will not block engineers currently until we are ready to turn it on

## Reviewer Notes
- left inline comments but I will be removing the github action once we're comfortable about the PR since we are not going to block developers yet
- will also turn off the no-unapproved-annotations rule once PR is done being tested

## Setup
`yarn install`
`yarn run lint:production` will run the rules on the entire project
Should show some lint error messages: `Disabling of this rule requires an annotation ...`

## Code Review Verification Steps
* [ ] Request review from a member of a different team.
* [ ] Have the Jira acceptance criteria been met for this change?

## References

* [Jira story](https://dp3.atlassian.net/browse/MB-7489) for this change

### CODE REVIEW

File: eslint/no-unapproved-annotation.js [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+const validatorStatusOptions = new Set([
+  'RA Accepted',
+  'Return to Developer',
+  'Known Issue',
+  'Mitigated',
+  'False Positive',
+  'Bad Practice',
+]);
+
+const approvedBypassableRules = new Set([
```
Comment (Ronolibert) [inline eslint/no-unapproved-annotation.js:21]: In the spreadsheet, `security/detect-object-injection` was originally a CAT I but it was deemed ok, should we add this rule to the list of rules that are ok to bypass?

File: eslint/no-unapproved-annotation.js [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+const getValidatorStatus = (commentsArr) =>
+  commentsArr.reduce((accum, curr) => {
+    if (curr.startsWith(VALIDATOR_LABEL)) {
+      // eg. RA Validator Status: Mitigated
+      return curr.split(':')[1].trim();
+    }
+    return accum;
+  }, '');
+
+/*
```
Comment (Ronolibert) [inline eslint/no-unapproved-annotation.js:56]: These are then scenarios that the linter flagged that were false positives. Theres only 12 instances but I didn't want to fix it in this PR

File: .eslintrc.js [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+    'prettier/prettier',
+    'plugin:security/recommended',
+    'plugin:you-dont-need-lodash-underscore/compatible',
+  ],
+  root: true,
+  rules: {
+    'prettier/prettier': 'error',
+    'no-only-tests/no-only-tests': 'error',
+    'cato/no-explicit-eslint-disable': 'error',
+    'cato/no-unapproved-annotation': process.env.NODE_ENV === 'production' ? 'error' : 'off',
```
Comment (Ronolibert) [inline .eslintrc.js:16]: Just leaving this in here for what the use case will be, but before we merge I'll turn this lint rule off so that it doesn't block people yet

File: .github/workflows/lint.yml [hunk 4]
```diff
@@ -0,0 +1,23 @@
+name: Lint
```
Comment (Ronolibert) [inline .github/workflows/lint.yml:1]: Just leaving this in here so that theres a full picture of how the linter will be used once it is ready to be turned on. But, before we merge I will remove this file

File: src/pages/Admin/Users/UserShow.jsx [hunk 5]
```diff
@@ -20,7 +20,7 @@ UserShowTitle.defaultProps = {
 
 const UserShow = (props) => {
   return (
-    // eslint-disable-next-line
+    // eslint-disable-next-line react/jsx-props-no-spreading
```
Comment (Ronolibert) [inline src/pages/Admin/Users/UserShow.jsx:23]: I fixed this because it only instance of a blocking change

**Review by kimallen — COMMENTED** [review]
This is really cool!  It's so nice to see exactly where all the errors are and where we need annotations or approvals.

A few inline comments.  Will wait for approval for when eslintrc is in line with master.  You should probably get some other eyes on this too before merging.

File: eslint/no-unapproved-annotation.js (hunk 1 above)
Comment (kimallen) [reply eslint/no-unapproved-annotation.js:21]: Not sure.  I wish Jarod was still around to ask about that.

File: eslint/no-unapproved-annotation.js [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+            loc: {
+              start: {
+                ...comment.loc.start,
+                column: -1,
+              },
+              end: comment.loc.end,
+            },
+            messageId: NO_ANNOTATION_MESSAGE_ID,
+          });
+        } else if (!approvedBypassableRules.has(rule) && !validatorStatusOptions.has(validatorStatus)) {
```
[the remark this answers is not in the document]
Comment (Ronolibert) [reply eslint/no-unapproved-annotation.js:126]: Done! I purposely didn't originally because it'd allow for weird casings but I doubt the validator would have statuses like `mItGateD

File: .github/workflows/lint.yml (hunk 4 above)
Comment (kimallen) [inline .github/workflows/lint.yml]: So this file mean that we will run eslint in Github Actions instead of CircleCI?  So when we implement we will remove eslint from the `circleci/config.yaml`?

File: tests/no-unapproved-annotation.js [hunk 7]
```diff
@@ -0,0 +1,81 @@
+const { RuleTester } = require('eslint');
+
+const rule = require('../eslint/no-unapproved-annotation');
+
+const ruleTester = new RuleTester();
+
+ruleTester.run('no-unapproved-annotation', rule, {
+  valid: [
```
[the remark this answers is not in the document]
Comment (Ronolibert) [reply tests/no-unapproved-annotation.js]: So this is one of the edge cases, the `getCommentsBefore` api can't find the annotation because the code is parsed differently when eslint-disable-line is used since its a mix of both real code and comments. Right now, there's only 1 instance in the code base like this

File: package.json [hunk 8]
```diff
@@ -133,6 +136,7 @@
     "enzyme": "^3.11.0",
     "eslint-config-airbnb": "^18.2.1",
     "eslint-config-prettier": "^8.1.0",
+    "eslint-plugin-cato": "file:./eslint",
```
[the remark this answers is not in the document]
Comment (Ronolibert) [reply package.json:138]: I'll change it to ato!

File: .github/workflows/lint.yml (hunk 4 above)
Comment (Ronolibert) [reply .github/workflows/lint.yml]: Yup! We could remove it from config.yaml since it'll just be running the same linter twice

File: .eslintrc.js (hunk 3 above)
Comment (Ronolibert) [reply .eslintrc.js:16]: I'm not sure what keyword to use here but it can be anything. In this case, I'm using production to mean anything not local. So once it hits the PR step its technically production. I could change this to be the opposite. I could have a `development` environment variable and say if it's not development then it'd be an error. Would that make more sense?

Comment (kimallen) [reply .eslintrc.js]: Hmm, what about 'deployed' or 'deployedenvt'.  Does it run on experimental?

Comment (kimallen) [reply .eslintrc.js]: though yes, re-reading your response, I think development might make more sense, esp if it runs for experimental.  Open to others to weigh-in too

Comment (Ronolibert) [reply .eslintrc.js]: We could include this in the WIP PR that will be prepped for when we turn this on

Comment (suzubara) [reply .eslintrc.js]: `process.env.NODE_ENV` is a built-in value that cannot be changed arbitrarily. from the [Create React App docs](https://create-react-app.dev/docs/adding-custom-environment-variables/):
> There is also a built-in environment variable called NODE_ENV. You can read it from process.env.NODE_ENV. When you run npm start, it is always equal to 'development', when you run npm test it is always equal to 'test', and when you run npm run build to make a production bundle, it is always equal to 'production'. You cannot override NODE_ENV manually. This prevents developers from accidentally deploying a slow development build to production.

so in the context of CRA, running webpack locally ie with `make client_run` will always be development, and the deployed sites that use `make client_build` will always be production.

Comment (suzubara) [reply .eslintrc.js]: I don't know the value of this when running the linter but it could be set based on the environment where it's run (ie, circleCI)

File: .eslintrc.js [hunk 9]
```diff
@@ -0,0 +1,88 @@
+module.exports = {
```
Comment (suzubara) [inline .eslintrc.js:1]: minor suggestion - if you `git mv .eslintrc .eslintrc.js` the diff should be easier to parse instead of when removing/adding a file

File: eslint/package.json [hunk 10]
```diff
@@ -0,0 +1,5 @@
+{
+  "name": "eslint-plugin-ato",
```
Comment (suzubara) [inline eslint/package.json:2]: since this `eslint` directory is specifically for the plugin, I would suggest renaming it to `eslint-plugin-ato`?

File: tests/no-explicit-eslint-disable.test.js [hunk 11]
```diff
@@ -0,0 +1,34 @@
+const { RuleTester } = require('eslint');
```
Comment (suzubara) [inline tests/no-explicit-eslint-disable.test.js:1]: since these tests are specific to the plugin, could they go inside `./eslint-plugin-ato/tests`?

**Review by suzubara — COMMENTED** [review]
This is really cool!! I left some comments around file naming that I think would help keep things organized. I also turned the rule on locally to see it show up in VSCode: 
image

very cool!! 🎉

File: .eslintrc.js (hunk 9 above)
Comment (Ronolibert) [reply .eslintrc.js:1]: Oooh, that makes sense! I'll do that to tidy up the PR slightly

File: eslint/package.json (hunk 10 above)
Comment (Ronolibert) [reply eslint/package.json:2]: I can do that!

**Comment by Ronolibert** [comment]
@suzubara Using git mv didn't appear to clean up the PR, but I renamed the lint folder and colocated the tests!

File: .eslintrc.js (hunk 9 above)
Comment (suzubara) [reply .eslintrc.js:1]: welp i don't know why that didn't work, git is a mystery 🙃

**Review by suzubara — APPROVED** [review]

File: eslint-plugin-ato/tests/no-unapproved-annotation.test.js [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+    },
+    {
+      code: '// eslint-disable security/detect-unsafe-regex',
+      errors: [
+        {
+          message:
+            'Disabling of this rule requires an annotation. Please visit https://docs.google.com/document/d/1qiBNHlctSby0RZeaPzb-afVxAdA9vlrrQgce00zjDww/edit?usp=sharing',
+        },
+      ],
+    },
```
Comment (felipe-lee) [inline eslint-plugin-ato/tests/no-unapproved-annotation.test.js:79]: Could you add an example of a rule with an `@` sign in it.

**Review by felipe-lee — COMMENTED** [review]
This is great! My comments are mainly suggestions to expand the things shown in the tests, but aren't required

File: .eslintrc.js [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+    'prettier/prettier',
+    'plugin:security/recommended',
+    'plugin:you-dont-need-lodash-underscore/compatible',
+  ],
+  root: true,
+  rules: {
+    'prettier/prettier': 'error',
+    'no-only-tests/no-only-tests': 'error',
+    'ato/no-explicit-eslint-disable': 'error',
+    'ato/no-unapproved-annotation': process.env.NODE_ENV === 'production' ? 'error' : 'off',
```
Comment (gidjin) [inline .eslintrc.js:16]: Just curious, is this the line that would change to turn this linter on?

File: eslint-plugin-ato/package.json [hunk 14]
```diff
@@ -0,0 +1,5 @@
+{
+  "name": "eslint-plugin-ato",
+  "version": "1.0.0",
+  "main": "index.js"
+}
```
Comment (gidjin) [inline eslint-plugin-ato/package.json:5]: This plugin is great! I'm wondering how much work it would be to make it an npm package vs in the mymove repo. I can see it being of value to other projects. It wouldn't need to be more generic initially, imo.

**Review by gidjin — APPROVED** [review]
This looks good to me. Thanks for putting this together. I left a couple questions but I'm good with approving. 🚀

**Comment by gidjin** [comment]
FWIW I realized after running the commands in the setup section I had a modified `yarn.lock`. Not sure why, but you may need to double check that before merging.

File: eslint-plugin-ato/package.json (hunk 14 above)
Comment (suzubara) [reply eslint-plugin-ato/package.json:5]: yes I agree :D

File: package.json [hunk 15]
```diff
@@ -108,6 +110,7 @@
   "version": "0.1.0",
   "devDependencies": {
     "@babel/core": "~7.13.14",
+    "@babel/helper-builder-react-jsx": "^7.12.13",
```
Comment (suzubara) [inline package.json:113]: what is this package for?
